lib/AST/ReachBasedRefine.cpp

```diff
@@ -40,83 +24,49 @@ z3::expr ReachBasedRefine::GetZ3Cond(clang::IfStmt *ifstmt) {
-    return Prove(*z3_ctx, z3::mk_or(conds) == z3_ctx->bool_val(true));
-  };
-
-  // Gather else-if candidates
-  for (auto stmt : llvm::make_range(stmts.rbegin(), stmts.rend())) {
-    // Quit if we gathered enough IfStmts for a cascade.
-    // This is recognized when the conjuction of reaching
-    // conditions of all the IfStmts form a tautology.
-    if (IsTautology()) {
-      break;
-    }
-    // Clear else-if IfStmts if we find a path among them.
-    auto cond = GetZ3Cond(stmt);
-    if (stmt->getElse() || !IsUnrechable(cond)) {
-      conds = z3::expr_vector(*z3_ctx);
-      elifs.clear();
-    }
-    // Add the current if-statement to the else-if candidates.
-    conds.push_back(cond);
-    elifs.push_back(stmt);
-  }
-
-  // Check if we have enough statements to work with
-  if (elifs.size() < 2) {
-    return;
-  }
-
-  // Create the else-if cascade
-  clang::IfStmt *sub = nullptr;
-  for (auto stmt : llvm::make_range(elifs.rbegin(), elifs.rend())) {
-    auto cond = stmt->getCond();
-    auto then = stmt->getThen();
-    if (stmt == elifs.back()) {
-      sub = ast.CreateIf(cond, then);
-      substitutions[stmt] = sub;
-    } else if (stmt == elifs.front()) {
-      std::vector<clang::Stmt *> thens({then});
-      sub->setElse(ast.CreateCompoundStmt(thens));
-      substitutions[stmt] = nullptr;
-    } else {
-      auto elif = ast.CreateIf(cond, then);
-      sub->setElse(elif);
-      sub = elif;
-      substitutions[stmt] = nullptr;
-    }
-  }
-}
-
 bool ReachBasedRefine::VisitCompoundStmt(clang::CompoundStmt *compound) {
-  // DLOG(INFO) << "VisitCompoundStmt";
-  // Create else-if cascade substitutions for IfStmts in `compound`
-  CreateIfElseStmts(GetIfStmts(compound));
-  // Apply created else-if substitutions and
-  // create a replacement for `compound`
-  if (ReplaceChildren(compound, substitutions)) {
-    std::vector<clang::Stmt *> new_body;
-    for (auto stmt : compound->body()) {
-      if (stmt) {
-        new_body.push_back(stmt);
+  std::vector<clang::Stmt *> body{compound->body_begin(), compound->body_end()};
+  std::vector<clang::IfStmt *> ifs;
+  z3::expr_vector conds{*z3_ctx};
+  bool done_something{false};
+  for (size_t i{0}; i < body.size(); ++i) {
+    if (auto if_stmt = clang::dyn_cast<clang::IfStmt>(body[i])) {
+      ifs.push_back(if_stmt);
+      auto cond{GetZ3Cond(if_stmt)};
+      if (!if_stmt->getElse() && Prove(*z3_ctx, !(cond && z3::mk_or(conds)))) {
+        conds.push_back(GetZ3Cond(if_stmt));
+
+        if (Prove(*z3_ctx, z3::mk_or(conds)) && ifs.size() > 2) {
+          auto last_if{ifs[0]};
+          for (auto stmt : ifs) {
+            if (stmt == ifs.front()) {
+              continue;
+            }
+            if (stmt == ifs.back()) {
+              last_if->setElse(stmt->getThen());
+            } else {
+              last_if->setElse(stmt);
+              last_if = stmt;
+            }
+          }
+
+          size_t start_delete{i - (ifs.size() - 2)};
```

Comment:
Same as above. This one removes the `if` statements that were put in the cascade, right?

Reply:
Ditto